pkg/hs/urbit-king/lib/Urbit/Vere/Ports.hs

```diff
@@ -0,0 +1,232 @@
+                filteredHeap
+          atomically notifyComplete
+          loop pmp withRenew
+
+    PTMRequestClose p -> do
+      logInfo $
+        displayShow ("port: releasing lease for ", p)
+      setPortMapping pmp PTUdp p p 0
+      let removed = filterPort p nextRenew
+      loop pmp removed
+
+  handleRenew :: NatPmpHandle
+              -> MinPrioHeap POSIXTime RenewAction
+              -> RIO e ()
+  handleRenew pmp nextRenew = do
+    case (Data.Heap.view nextRenew) of
+      Nothing -> error "Internal heap managing error."
+      Just ((_, RenewAction p), rest) -> do
+        logInfo $
+          displayShow ("port: sending renewing request to NAT-PMP for port ",
+                       p)
+        setPortMapping pmp PTUdp p p portLeaseLifetime >>= \case
+          Left err -> do
+            logError $
+              displayShow ("port: failed to request NAT-PMP for port ", p,
+                           ":", err, ", disabling NAT-PMP")
+            loopErr q
+          Right _ -> do
+            -- We don't need to filter the port because we just did.
+            now <- io $ getPOSIXTime
+            let withRenew =
+                  insert (now + (fromIntegral portRenewalTime), RenewAction p)
+                  rest
+            loop pmp withRenew
+
+  filterPort :: Word16
+             -> MinPrioHeap POSIXTime RenewAction
+             -> MinPrioHeap POSIXTime RenewAction
+  filterPort p = Data.Heap.filter okPort
+    where
+      okPort (_, RenewAction x) = p /= x
+
+  -- block (retry) until the delay TVar is set to True
+  fini :: TVar Bool -> STM ()
+  fini = check <=< readTVar
+
+  -- The NAT system is considered "off" but we still need to signal back to
+  -- the main thread that blocking actions are complete.
+  loopErr q = forever $ do
+    (atomically $ readTQueue q) >>= \case
+      PTMRequestOpen _ onComplete -> atomically onComplete
+      PTMRequestClose _ -> pure ()
+
+-- When we were unable to connect to a router, get the ip address on the
+-- default ipv4 interface to check if we look like we're on an internal network
+-- or not.
+likelyIPAddress :: MonadIO m => m (Maybe (Word8, Word8, Word8, Word8))
+likelyIPAddress = liftIO do
+  -- Try opening a socket to 1.1.1.1 to get our own IP address. Since UDP is
+  -- stateless and we aren't sending anything, we aren't actually contacting
+  -- them in any way.
+  sock <- socket AF_INET Datagram 0
+  connect sock (SockAddrInet 53 (tupleToHostAddress (1, 1, 1, 1)))
+  sockAddr <- getSocketName sock
+  case sockAddr of
+    SockAddrInet _ addr -> pure $ Just $ hostAddressToTuple addr
+    _                   -> pure $ Nothing
+
+-- Acquire a port for the duration of the RAcquire.
+requestPortAccess :: forall e. (HasPortControlApi e) => Word16 -> RAcquire e ()
+requestPortAccess port = do
+  mkRAcquire request release
+ where
+  request :: RIO e ()
+  request = do
+    api <- asks (^. portControlApiL)
```

Comment:
Very sorry. The idomatic way to use a lens on the reader environment is in fact

```
view lens
```
not
```
asks (^. lens)

Reply:
Changed all asks to view.